### PR TITLE
add unit test for vineyardctl

### LINKED ISSUE #1301: Add unit test for vineyardctl and vineyard operator
Describe your problem
---------------------

At present, we already have the e2e tests for every important feature of the vineyard operator and the vineyardctl. However, the unit tests for [vineyardctl](https://github.com/v6d-io/v6d/tree/main/k8s/cmd), [controllers](https://github.com/v6d-io/v6d/tree/main/k8s/controllers/k8s) and [webhooks](https://github.com/v6d-io/v6d/tree/main/k8s/pkg/webhook)  are lacking, so the issue mainly contains the following contents.

- [ ] Complete the unit test for [vineyardctl](https://github.com/v6d-io/v6d/tree/main/k8s/cmd) refer to the [doc of writing golang test case](https://go.dev/doc/tutorial/add-a-test).
- [ ] Complete the unit test for [controllers](https://github.com/v6d-io/v6d/tree/main/k8s/controllers/k8s) refer to the [doc of writing controllers test case](https://book.kubebuilder.io/cronjob-tutorial/writing-tests.html)
- [ ] Complete the unit test for [webhooks](https://github.com/v6d-io/v6d/tree/main/k8s/pkg/webhook) refer to the [doc of writing golang test case](https://go.dev/doc/tutorial/add-a-test).

### PR DESCRIPTION
**Pull request by zhuyi1159**
What do these changes do?
-------------------------

add unit test for vineyardctl

Related issue number
--------------------

Fixes #1301

### CODE REVIEW

File: k8s/cmd/commands/sidecar/inject_test.go [hunk 1]
```diff
@@ -0,0 +1,1057 @@
...
+	"github.com/v6d-io/v6d/k8s/cmd/commands/flags"
+	"github.com/v6d-io/v6d/k8s/cmd/commands/util"
+	"github.com/v6d-io/v6d/k8s/controllers/k8s"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestInjectCmd(t *testing.T) {
+	testReplicas := struct {
```
Comment (dashanji) [inline k8s/cmd/commands/sidecar/inject_test.go:42]: Update the test name here.

File: k8s/cmd/commands/sidecar/inject_test.go [hunk 2]
```diff
@@ -0,0 +1,1057 @@
...
+		for _, svc := range svcList.Items {
+			if svc.Spec.Ports[0].Port != int32(testReplicas.expectedService_port) {
+				t.Errorf("Service %s in namespace %s uses the port %d, expected port %d\n", svc.Name, svc.Namespace, svc.Spec.Ports[0].Port, testReplicas.expectedService_port)
+			}
+			if string(svc.Spec.Type) != testReplicas.expectedService_type {
+				t.Errorf("Service %s in namespace %s uses the type %s, expected type %s\n", svc.Name, svc.Namespace, string(svc.Spec.Type), testReplicas.expectedService_type)
+			}
+		}
+
+		// 获取 PVC 对象
```
Comment (dashanji) [inline k8s/cmd/commands/sidecar/inject_test.go:192]: There is no need to test pvc here.

File: k8s/cmd/commands/sidecar/inject_test.go [hunk 3]
```diff
@@ -0,0 +1,1057 @@
...
+			},
+		}
+		err = k8sclient.List(context.Background(), &etcdPod, etcdOpts...)
+		if err != nil {
+			t.Errorf("list etcd pods error: %v", err)
+		}
+
+		for _, pod := range vineyardPods.Items {
+			for _, container := range pod.Spec.Containers {
+				if container.Image != testReplicas.expectedImage {
```
Comment (dashanji) [inline k8s/cmd/commands/sidecar/inject_test.go:137]: It's a bug here. Please fix it as the container of the injected workload is not `vineyardcloudnative/vineyardd:alpine-latest`. Also, could you please check the other parts at the same time?

File: k8s/cmd/commands/sidecar/inject_test.go [hunk 4]
```diff
@@ -0,0 +1,1057 @@
...
+		workloadYAML   string
+		workloadJSON   string
+		expectedResult string
+		expectedError  error
+	}{
+		{
+			name:         "Valid YAML file",
+			workloadYAML: "/home/zhuyi/v6d/k8s/config/samples/k8s_v1alpha1_sidecar.yaml",
+			workloadJSON: "",
+			expectedResult: "apiVersion: k8s.v6d.io/v1alpha1\n" +
```
Comment (dashanji) [inline k8s/cmd/commands/sidecar/inject_test.go:253]: Please use the single quotation.

File: k8s/cmd/commands/sidecar/inject_test.go [hunk 5]
```diff
@@ -0,0 +1,1057 @@
...
...{"metadata":{"creationTimestamp":null,"labels":{"app":"nginx","app.vineyard.io/name":"vineyard-sidecar","app.vineyard.io/role":"vineyardd"}},"spec":{"containers":[{"command":null,"image":"nginx:1.14.2","name":"nginx","ports":[{"containerPort":80}],"resources":{},"volumeMounts":[{"mountPath":"/var/run","name":"vineyard-socket"}]},{"command":["/bin/bash","-c","/usr/bin/wait-for-it.sh -t 60 vineyard-sidecar-etcd-service.vineyard-system.svc.cluster.local:2379; sleep 1; /usr/local/bin/vineyardd --sync_crds true --socket /var/run/vineyard.sock --size  --stream_threshold 80 --etcd_cmd etcd --etcd_prefix /vineyard --etcd_endpoint http://vineyard-sidecar-etcd-service:2379\n"],"env":[{"name":"VINEYARDD_UID","value":null},{"name":"VINEYARDD_NAME","value":"vineyard-sidecar"},{"name":"VINEYARDD_NAMESPACE","value":"vineyard-system"}],"image":"vineyardcloudnative/vineyardd:latest","imagePullPolicy":"IfNotPresent","name":"vineyard-sidecar","ports":[{"containerPort":9600,"name":"vineyard-rpc","protocol":"TCP"}],"resources":{"limits":null,"requests":null},"volumeMounts":[{"mountPath":"/var/run","name":"vineyard-socket"}]}],"volumes":[{"emptyDir":{},"name":"vineyard-socket"}]}}},"status":{}}` + "\n",
```
Comment (dashanji) [inline k8s/cmd/commands/sidecar/inject_test.go:433]: It's hard to maintain the long string, a possible way is to output the string as YAML.

File: k8s/cmd/commands/sidecar/inject_test.go [hunk 6]
```diff
@@ -0,0 +1,1057 @@
...
+
+	// Mock ConvertToYaml function
+	//util.ConvertToYaml = mockConvertToYaml
+
+	yamlManifests, err := parseManifestsAsYAML(om)
+	assert.NoError(t, err)
+	assert.Equal(t, expectedYAML, yamlManifests)
+}*/
+
+func Test_parseManifestsAsYAML(t *testing.T) {
```
Comment (dashanji) [inline k8s/cmd/commands/sidecar/inject_test.go:494]: The test looks very strange, it's okay to delete this.

File: k8s/cmd/commands/sidecar/inject_test.go [hunk 7]
```diff
@@ -0,0 +1,1057 @@
...
...p":"nginx"}},"strategy":{},"template":{"metadata":{"creationTimestamp":null,"labels":{"app":"nginx","app.vineyard.io/name":"vineyard-sidecar"}},"spec":{"containers":[{"command":null,"image":"nginx:1.14.2","name":"nginx","ports":[{"containerPort":80}],"resources":{},"volumeMounts":[{"mountPath":"/var/run","name":"vineyard-socket"}]},{"command":["/bin/bash","-c","/usr/bin/wait-for-it.sh -t 60 vineyard-sidecar-etcd-service.vineyard-system.svc.cluster.local:2379; sleep 1; /usr/local/bin/vineyardd --sync_crds true --socket /var/run/vineyard.sock --size 256Mi --stream_threshold 80 --etcd_cmd etcd --etcd_prefix /vineyard --etcd_endpoint http://vineyard-sidecar-etcd-service:2379\n"],"env":[{"name":"VINEYARDD_UID","value":null},{"name":"VINEYARDD_NAME","value":"vineyard-sidecar"},{"name":"VINEYARDD_NAMESPACE","value":"vineyard-system"}],"image":"vineyardcloudnative/vineyardd:latest","imagePullPolicy":"IfNotPresent","name":"vineyard-sidecar","ports":[{"containerPort":9600,"name":"vineyard-rpc","protocol":"TCP"}],"resources":{"limits":null,"requests":null},"volumeMounts":[{"mountPath":"/var/run","name":"vineyard-socket"}]}],"volumes":[{"emptyDir":{},"name":"vineyard-socket"}]}}},"status":{}`,
```
Comment (dashanji) [inline k8s/cmd/commands/sidecar/inject_test.go:579]: Looks the same as before, please reuse the previous string here.

File: k8s/cmd/commands/util/format_test.go
Comment (dashanji) [inline k8s/cmd/commands/util/format_test.go:1]: No need to test here.

File: k8s/cmd/commands/util/usage/doc_test.go
Comment (dashanji) [inline k8s/cmd/commands/util/usage/doc_test.go:1]: No need here.

File: k8s/cmd/commands/util/usage/json_test.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	"encoding/json"
+	"reflect"
+	"testing"
+
+	"github.com/spf13/cobra"
+)
+
+func TestUsageJson(t *testing.T) {
+	// 创建一个命令
+	cmd := &cobra.Command{
```
Comment (dashanji) [inline k8s/cmd/commands/util/usage/json_test.go:14]: It is not advisable to create a new command for testing.

File: k8s/cmd/commands/util/usage/md_docs_test.go
Comment (dashanji) [inline k8s/cmd/commands/util/usage/md_docs_test.go:1]: No need here.

File: k8s/cmd/commands/util/usage/types_test.go
Comment (dashanji) [inline k8s/cmd/commands/util/usage/types_test.go:1]: No need here.

File: k8s/cmd/commands/util/template_test.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+		},
+	}
+
+	objs, err := BuildObjsFromVineyarddManifests(files, value, tmplFunc)
+
+	if err != nil {
+		t.Fatalf("Failed to build objects from Vineyardd manifests: %v", err)
+	}
+
+	if len(objs) != 0 {
```
Comment (dashanji) [inline k8s/cmd/commands/util/template_test.go:114]: Please specify the number here.

File: k8s/cmd/commands/util/template_test.go [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+	}
+	vineyardd := value
+	name := vineyardd.Name
+	namespace := vineyardd.Namespace
+	replicas := vineyardd.Spec.EtcdReplicas
+	image := vineyardd.Spec.Vineyard.Image
+
+	_, _, err := BuildObjsFromEtcdManifests(&etcdConfig, name, namespace, replicas, image, value, tmplFunc)
+
+	assert.NoError(t, err)
```
Comment (dashanji) [inline k8s/cmd/commands/util/template_test.go:84]: Same as before.

File: k8s/cmd/commands/util/template_test.go [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+			return q.String()
+		},
+		"getEtcdConfig": func() k8s.EtcdConfig {
+			return etcdConfig
+		},
+	}
+
+	obj, err := RenderManifestAsObj(path, value, tmplFunc)
+
+	assert.NoError(t, err)
```
Comment (dashanji) [inline k8s/cmd/commands/util/template_test.go:52]: Could you please make the test more meaningful?

**Review by dashanji — CHANGES_REQUESTED** [review]
Great job on the PR! I can see that you have made some improvements since the previous version. However, there are still a few points that need attention.

- I notice you rewritten the main part of a command to adapt the input and output of a specific function, it's okay to delete such test cases. Also, it's better to make the unit test more flexible.
- Could you please revise the code to make the unit test more meaningful?
- Please keep an eye out for redundant tests and review the comments accordingly.
- It's essential to double-check unused code and debug information before submitting your PR.
- Be sure to use English for annotations.

Please keep up the good work!

File: k8s/cmd/commands/deploy/deploy_vineyardd_test.go [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+				t.Errorf("BuildVineyardManifestFromInput() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestBuildVineyardManifestFromFile(t *testing.T) {
+	// set the flags
+	flags.Namespace = "vineyard-system"
+	flags.VineyarddFile = os.Getenv("HOME") + "/v6d/k8s/test/e2e/vineyardd.yaml"
```
Comment (dashanji) [inline k8s/cmd/commands/deploy/deploy_vineyardd_test.go:307]: Please DON'T to hard-code the path.

**Review by dashanji — CHANGES_REQUESTED** [review]
I roughly reviewed the previous files. There are several issues you need to address.
- Delete the duplicated test functions.
- Every test function is a complete execution unit, please check in your test function and make sure they can be run successfully independently.
- Update the hard-coded path.

Otherwise, please add a GitHub CI to test the unit test.

**Review by dashanji — CHANGES_REQUESTED** [review]
Overall LGTM! There are several hints for you.
- Format several JSON strings to make them easier to read.
- Reduce the duplications and reuse the common parts between unit test functions. 
- Simplify these makefile parts.

File: k8s/test/e2e/Makefile [hunk 13]
```diff
@@ -394,6 +394,31 @@ e2e-tests-airflow-integration: prepare-e2e-test build-local-cluster load-vineyar
...
+# schedule test
+	@export KUBECONFIG="/tmp/e2e-k8s.config" && cd ${K8S_DIR} && go test ./cmd/commands/schedule/
+# sidecar test 
+	@export KUBECONFIG="/tmp/e2e-k8s.config" && cd ${CMD_DIR}/commands/sidecar && \
+		go test -run TestInjectCmd && go test -run Test_deployDuringInjection && go test -run slow
+# util test
+	@export KUBECONFIG="/tmp/e2e-k8s.config" && cd ${K8S_DIR} && go test ./cmd/commands/util/
+	@echo "vineyardctl unit test passed."
+	@make delete-local-cluster
+.PHONY: unittest
```
Comment (sighingnow) [inline k8s/test/e2e/Makefile:420]:
```suggestion
############# unit testing ################################################
unittest: prepare-e2e-test build-local-cluster load-vineyardd-image
	@make -C ${K8S_DIR} install-cert-manager
	@make -C ${E2E_DIR} install-vineyard-operator
	@echo "Running vineyardctl unit test.." 
    # create test
	@export KUBECONFIG="/tmp/e2e-k8s.config" && cd ${K8S_DIR} && go test ./cmd/commands/create/
   # delete test 
	@export KUBECONFIG="/tmp/e2e-k8s.config" && cd ${K8S_DIR} && go test ./cmd/commands/delete/
  # deploy test
	@export KUBECONFIG="/tmp/e2e-k8s.config" && cd ${CMD_DIR}/commands/deploy && \
		go test -run first && go test -run second && go test -run third 
    # schedule test
	@export KUBECONFIG="/tmp/e2e-k8s.config" && cd ${K8S_DIR} && go test ./cmd/commands/schedule/
    # sidecar test 
	@export KUBECONFIG="/tmp/e2e-k8s.config" && cd ${CMD_DIR}/commands/sidecar && \
		go test -run TestInjectCmd && go test -run Test_deployDuringInjection && go test -run slow
    # util test
	@export KUBECONFIG="/tmp/e2e-k8s.config" && cd ${K8S_DIR} && go test ./cmd/commands/util/
	@echo "vineyardctl unit test passed."
	@make delete-local-cluster
.PHONY: unittest
```

**Comment by sighingnow** [comment]
Merged. @zhuyi1159  Thanks a lot for your effort!

**Pull request merged**